06-classification_continued.Rmd

```diff
@@ -559,16 +657,378 @@ The overall workflow for performing K-nearest neighbour classification using `ti
+> accuracy on the test data and other future observations.
+> Since forward selection involves training a lot of models, you run a fairly
+> high risk of this happening. To keep this risk low, only use forward selection
+> when you have a large amount of data and a relatively small total number of 
+> predictors. More advanced methods do not suffer from this
+> problem as much; see the additional resources at the end of this chapter for
+> where to learn more about advanced predictor selection methods. 
+
+**Forward selection in R**
+ 
+We now turn to implementing forward selection in R.
+Unfortunately there is no built-in way to do this using the `tidymodels` framework,
+so we will have to code it ourselves. First we will use the `select` function
+to extract the "total" set of predictors that we are willing to work with. 
+Here we will load the modified version of the cancer data with irrelevant 
+predictors, and select `Smoothness`, `Concavity`, `Perimeter`, `Irrelevant1`, `Irrelevant2`, and `Irrelevant3`
+as potential predictors, and the `Class` variable as the response.
+We will also extract the column names for the full set of predictor variables.
+```{r 06-fwdsel, warning = FALSE}
+set.seed(1)
+cancer_subset <- cancer_irrelevant %>% select(Class, Smoothness, Concavity, Perimeter, Irrelevant1, Irrelevant2, Irrelevant3)
+names <- colnames(cancer_subset %>% select(-Class))
+
+cancer_subset
+```
+The key idea of the forward selection code is to use the `paste` function (which concatenates strings
+separated by spaces) to create a model formula for each subset of predictors for which we want to build a model.
+The `collapse` argument tells `paste` what to put between the items in the list;
+to make a formula, we need to put a `+` symbol between each variable.
+As an example, let's make a model formula for all the predictors,
+which should output something like
+`Class ~ Smoothness + Concavity + Perimeter + Irrelevant1 + Irrelevant2 + Irrelevant3`:
+```{r}
+example_formula <- paste("Class", "~", paste(names, collapse="+"))
+example_formula
+```
+
+Finally, we need to write some code that performs the task of sequentially
+finding the best predictor to add to the model.
+If you recall the end of the wrangling chapter, we mentioned
+that sometimes one needs more flexible forms of iteration than what 
+we have used earlier, and in these cases one typically resorts to
+[a for loop](https://r4ds.had.co.nz/iteration.html#iteration).
+This is one of those cases! Here we will use two for loops:
+one over increasing predictor set sizes 
+(where you see `for (i in 1:length(names))` below),
+and another to check which predictor to add in each round (where you see `for (j in 1:length(names))` below).
+For each set of predictors to try, we construct a model formula,
+pass it into a `recipe`, build a `workflow` that tunes
+a K-NN classifier using 5-fold cross-validation, 
+and finally records the estimated accuracy.
+
+```{r 06-fwdsel-2, warning = FALSE}
```

Comment:
We definitely need to tidy this, or abstract to a function in a package, I am thinking of a function with this spec:

```
add_variable_selection <- function(workflow, method = "forward", metric = "accuracy") {
  # return tibble with model objects, model formula and accuracies
}
```

What do you think?

Reply:
@ttimbers sounds good

I'm not going to wait until this happens to merge the PR. You can make a separate PR for that when you're done.